filestream: derive scanner fingerprinting from the file identity (#52147)
elastic/beats: docs/reference/filebeat/filebeat-input-filestream.md

Follow [this comprehensive guide](/reference/filebeat/file-identity.md) on how to choose a file identity option right for your use-case.

::::{applies-switch}
:::{applies-item} stack: ga 9.6+
:sync: 9.6+
Scanner fingerprinting follows the configured `file_identity`: Filebeat enables it for the `fingerprint` file identity (the default) and disables it for any other file identity. The `prospector.scanner.fingerprint.enabled` setting is deprecated and ignored.
:::
:::{applies-item} stack: ga 9.0-9.5
:sync: 9.0-9.5
Scanner fingerprinting is enabled by default. Set `prospector.scanner.fingerprint.enabled: false` when you configure any other file identity.
:::
:::{applies-item} stack: ga 8.10-8.19
:sync: 8.10-8.19
Scanner fingerprinting is disabled by default. Set `prospector.scanner.fingerprint.enabled: true` when you configure the `fingerprint` file identity.

> :::

Reviewer: The 8.x docs are in the 8.19 branch, in a totally different system. Those docs 'only' cover 9.x
```suggestion
```

Follow [this comprehensive guide](/reference/filebeat/file-identity.md) on how to choose a file identity option right for your use-case.

> Scanner fingerprinting follows the configured `file_identity`: Filebeat enables it for the `fingerprint` file identity (the default when `file_identity` is omitted) and disables it for any other file identity. The `prospector.scanner.fingerprint.enabled` setting is deprecated and ignored; Filebeat logs a warning if its value contradicts the configured `file_identity`.

Reviewer 2: ```suggestion
Scanner fingerprinting follows the configured `file_identity`: Filebeat enables it for the `fingerprint` file identity (the default when `file_identity` is omitted) and disables it for any other file identity. The `prospector.scanner.fingerprint.enabled` setting is deprecated and ignored. Filebeat logs a warning if its value contradicts the configured `file_identity`.
```

> Scanner fingerprinting follows the configured `file_identity`: Filebeat enables it for the `fingerprint` file identity (the default when `file_identity` is omitted) and deactivates it for any other file identity. The `prospector.scanner.fingerprint.enabled` setting is deprecated and ignored. Filebeat logs a warning if its value contradicts the configured `file_identity`.

Reviewer: This section needs to explain both behaviours: before this PR and after, with the specific versions either mentioned in the text or using the `applies_to` tag. Users need to know exactly with versions require setting `prospector.scanner.fingerprint.enabled` and which ones do not.

:   The default behavior of Filebeat is to identify files based on content by hashing a specific range (0 to 1024 bytes by default).

::::{warning}

> This file identity option uses file fingerprints produced by the [scanner](#filebeat-input-filestream-scan-fingerprint), which is enabled automatically when this file identity is used. Once this file identity is enabled, changing the fingerprint configuration (offset, length, or other settings) will lead to a global re-ingestion of all files that match the paths configuration of the input.

Reviewer: Same thing here: both behaviours need to be explained with with the versions they apply to.

$$$filebeat-input-filestream-file-identity-fingerprint$$$

**`fingerprint`**
:   Identifies files based on content by hashing a specific range (0 to 1024 bytes by default).

This file identity option uses file fingerprints produced by the [scanner](#filebeat-input-filestream-scan-fingerprint).

::::{applies-switch}
:group: fingerprint-scanner
:::{applies-item} stack: ga 9.5+
:sync: 9.5+
Scanner fingerprinting is enabled automatically when using this identity.
:::
:::{applies-item} stack: ga 9.0-9.5
:sync: 9.0-9.5
Scanner fingerprinting is enabled by default. If you explicitly set `prospector.scanner.fingerprint.enabled: false`, this file identity will not work.
:::
:::{applies-item} stack: ga 8.10-8.19
:sync: 8.10-8.19
Scanner fingerprinting is disabled by default. Set `prospector.scanner.fingerprint.enabled: true` to use this file identity.

> :::

Reviewer: Same thing here
```suggestion
```

    Depending on a mounting approach, the device ID (which is also used for comparing files) might change after a reboot.

> The scanner computes fingerprints when the [`fingerprint` file identity](#filebeat-input-filestream-file-identity-fingerprint) (the default) is used and skips them for any other file identity. The `prospector.scanner.fingerprint.enabled` setting is deprecated and ignored.

Reviewer: This paragraph also needs the `applies_to` tag.

    Depending on a mounting approach, the device ID (which is also used for comparing files) might change after a reboot.

**Configuration**

::::{warning}

> {applies_to}`stack: ga 8.10-9.4` Enabling fingerprint mode delays ingesting new files until they grow to at least `offset`+`length` bytes in size. Until then, these files are ignored.

Reviewer: ```suggestion
{applies_to}`stack: ga 9.0-9.4` Enabling fingerprint mode delays ingesting new files until they grow to at least `offset`+`length` bytes in size. Until then, these files are ignored.
```

{applies_to}`stack: ga 9.5+` Enhanced fingerprint tracks smaller files by default. The ingestion delay applies only when `file_identity.fingerprint.growing` is set to `false`.
::::

Normally, log lines contain timestamps and other unique fields that make the default fingerprint range suitable. Inspect your logs to determine appropriate `offset` and `length` values. The default `offset` is `0`, and the default `length` is `1024` bytes. `length` cannot be less than `64` bytes.

::::{applies-switch}
:group: fingerprint-scanner
:::{applies-item} stack: ga 9.0+
:sync: 9.0+
```yaml
fingerprint:
  offset: 0
  length: 1024
```
:::
:::{applies-item} stack: ga 8.10-8.19
:sync: 8.10-8.19
```yaml
fingerprint:
  enabled: true
  offset: 0
  length: 1024
```

> :::

Reviewer: No need for 8.19 docs
```suggestion
```
